pkg/ingester/instance.go

```diff
@@ -145,16 +148,32 @@ func (i *instance) Push(ctx context.Context, req *logproto.PushRequest) error {
 	defer recordPool.PutRecord(record)
 
 	var appendErr error
-	for _, s := range req.Streams {
-
-		err := i.executeWithGetOrCreateStream(s, record, func(stream *stream) error {
-			_, err := stream.Push(ctx, s.Entries, record, 0, false)
-			return err
-		})
+	for _, reqStream := range req.Streams {
+
+		s, _, err := i.streams.LoadOrStoreNew(reqStream.Labels,
+			func() (*stream, error) {
+				s, err := i.createStream(reqStream, record)
+				// Lock before adding to maps
+				if err == nil {
+					s.chunkMtx.Lock()
+				}
```

Comment:
Im' curious. Why must the lock happen before adding it to maps?

Reply:
This is to prevent a race condition against the flush code, which will also try to lock `cons mtx -> chunk mtx` before deleting a stream. If we don't lock the `chunkmtx` here, it's possible for the stream to be orphaned (deleted from streams map) in between creating it and pushing to it in this `Push()` call, resulting in data loss.

Reply:
To prevent this stream being deleted by another goroutine (periodic flush). So current goroutine can create a stream guaranteed in `streamsMap` until unlock later.

Reply:
Ok, so the other option would have been to keep a stream lock instead. This should also prevent a stream from being orphaned. If I'm not mistaken.